src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -66,26 +67,41 @@ public static string GetNuGetLockFilePath(string baseDirectory, string projectNa
         /// </summary>
         /// <param name="dgSpec">The <see cref="DependencyGraphSpec"/> for the new project defintion.</param>
         /// <param name="nuGetLockFile">The current <see cref="PackagesLockFile"/>.</param>
-        /// <returns>True if the lock file is valid false otherwise. </returns>
-        public static bool IsLockFileStillValid(DependencyGraphSpec dgSpec, PackagesLockFile nuGetLockFile)
+        /// <returns>True if the lock file is valid false otherwise.
+        /// The second return type is a localized message that indicates in further detail the reason for the inconsistency.</returns>
+        public static (bool, string) IsLockFileStillValid(DependencyGraphSpec dgSpec, PackagesLockFile nuGetLockFile)
```

Comment:
From a public API design point of view, I'm not a fan of using tuples. While the xmldoc has explained what the values mean, doing `var result = PackagesLockFileUtilities.IsLockFileStillValid(...)`, will cause `result` to have `Item1` and `Item2`, which are meaningless names.  I can't think off the top of my head any BCL class that returns value tuples. So, I think value tuples are great for private methods, so you can look up the method in the same file. I'm less enthusiastic across class boundaries (but still internal to the assembly), and think it maybe should be avoided across assembly boundaries.

Additionally, excluding cases when "this should never happen", the `bool` is always `true` when the `string` is the empty string, and every time the `string` is not empty, the `bool` is `false`. If we're confident about those "this should never happen" cases really never happening, they could throw exceptions, and we can get rid of the `bool` from the return value. This means we could return only the message, and the bool is redundant, as we can calculate it.

Also, I can imagine it might be useful in the future to be able to format the errors depending on the context that they're being displayed. Specifically, the lock file might have multiple differences between the restored project. It should be possible for callers of this API to put each reason on its own line. Therefore, I think returning a list of reasons, rather than a single message string, would be a better API design, as it lets the caller decide how the message should be displayed. I didn't look at the possible messages closely enough, but maybe a structured type would give more flexibility than a string, so that the package id is more easily determined programmatically, allowing a smart GUI to give users an easier way to resolve their problem.

Anyway, there are many possibilities for a public API, and as stated, I'm not a fan of tuples in public APIs. What do you think?

Reply:
+1. I agree with every point in above comment Andy. I am working on alternate design for this public API.

Reply:
I see that we already have a class named `LockFileValidityWithMatchedResults`  for `Packages Config lock file scenarios`.

https://github.com/NuGet/NuGet.Client/blob/b293bd5cac6ef6d88c315f75ba69cbe47db3227e/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs#L443-L446

I added a similar type that contains lock file validity status and list of invalid reasons which can be returned to the caller.

https://github.com/NuGet/NuGet.Client/blob/b66e164a3cea2bff2cbc4ae3ead72eb4dea7d0af/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs#L578-L581

Please review again and let me know your feedback.